pkg/controller/components/apiproxy.go

```diff
@@ -0,0 +1,166 @@
+package components
+
+import (
+	"fmt"
+
+	networkingv1 "k8s.io/api/networking/v1"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/dominodatalab/distributed-compute-operator/pkg/cluster/metadata"
+	"github.com/dominodatalab/distributed-compute-operator/pkg/controller/actions"
+	"github.com/dominodatalab/distributed-compute-operator/pkg/controller/core"
+)
+
+const (
+	defaultAPIProxyPort        = 8899
+	targetServicePort          = 8899
+	apiProxyPortName           = "api-proxy"
+	component                  = "api-proxy"
+	executionIDLabel           = "dominodatalab.com/execution-id"
+	projectIDLabel             = "dominodatalab.com/project-id"
+	datasourceProxyClientLabel = "datasource-proxy-client"
+)
+
+func executionID(obj *client.Object) string {
+	if executionID, ok := (*obj).GetLabels()[executionIDLabel]; ok {
+		return executionID
+	}
+	return "unknown-execution-id" // primarily, for integration testing
+}
+
+func runPodSelector(obj *client.Object) map[string]string {
+	runPodName := fmt.Sprintf("run-%s", executionID(obj))
+	return map[string]string{
+		metadata.ApplicationInstanceLabelKey: runPodName,
+	}
+}
+
+func newResourceMeta(obj *client.Object, componentMeta *metadata.Provider) metav1.ObjectMeta {
+	instanceName := fmt.Sprintf("%s-%s", component, executionID(obj))
+	return metav1.ObjectMeta{
+		Name:      instanceName,
+		Namespace: (*obj).GetNamespace(),
+		Labels:    componentMeta.StandardLabelsWithComponent(*obj, component, nil),
+	}
+}
+
+func NewAPIProxyServiceComponent(obj *client.Object, port int32, meta *metadata.Provider) *corev1.Service {
+	var apiProxyPort int32
+	if port != 0 {
+		apiProxyPort = port
+	} else {
+		apiProxyPort = defaultAPIProxyPort
+	}
+
+	ports := []corev1.ServicePort{{
+		Name:       apiProxyPortName,
+		Port:       apiProxyPort,
+		TargetPort: intstr.FromInt(targetServicePort),
```

Comment:
OK, you answered here my question from the first comment (about ports in CRD).

I suppose, Nucleus should then set the port on `jwt-tooling` container as well, based on the same value. (I believe now it's hard-coded).

Reply:
I'm not sure what you are referring to. Do we hardcode the tooling container here? That might need to be removed anyway since we are moving to API proxy.

Reply:
Let's say, I change the DCO code such that the CRD ports parameter applies to both listening port and the target port of the Service. Then, if we want Nucleus to set this (e.g., via a central config or any other external parameter), we also want it to simultaneously change which port the API proxy is listening to in its side-car container. Right now, it's probably just hardcoded to 8899.

Reply:
Yes, it may be hardcoded in Nucleus, I think my concerns is that you have the port that the Service is listening on, and then the target port on the Pod. Nucleus actually controls / specifies the Pod port. But it also makes sense to just set the Service port to match, since Nucleus will also specify the env var and needs to know what port the Service is listening on.

Reply:
Yes, I am fine with setting both ports in the Service to the same value.